File: frontend/src/types/template.ts
export type TemplateCategory = 'vlan' | 'interface' | 'acl' | 'ntp' | 'snmp' | 'custom'

export interface VariableDefinition {
  name: string
  type: 'string' | 'number' | 'boolean' | 'list'
  required: boolean
  default: string | null
  description: string | null
}

export interface Template {
  id: string
  name: string
  category: TemplateCategory
  description: string | null
  content: string
  variables: VariableDefinition[]
  created_by: string | null
  created_at: string
  updated_at: string
}

export interface TemplateCreate {
  name: string
  category: TemplateCategory
  description?: string | null
  content: string
  variables: VariableDefinition[]
}

export interface TemplateUpdate {
  name?: string
  category?: TemplateCategory
  description?: string | null
  content?: string
  variables?: VariableDefinition[]
}

export interface TemplateListResponse {
  items: Template[]
  total: number
}

export interface PreviewRequest {
  content: string
  variable_values: Record<string, string>
}

export interface TemplatePreviewRequest {
  variable_values: Record<string, string>
}

export interface PreviewResponse {
  rendered: string
  errors: string[]
  variables_found: string[]
}

File: backend/app/schemas/template.py
import uuid
from datetime import datetime
from typing import Literal
from pydantic import BaseModel, field_validator
from app.models.template import TemplateCategory


class VariableDefinition(BaseModel):
    name: str
    type: Literal["string", "number", "boolean", "list"] = "string"
    required: bool = True
    default: str | None = None
    description: str | None = None


class TemplateCreate(BaseModel):
    name: str
    category: TemplateCategory = TemplateCategory.custom
    description: str | None = None
    content: str
    variables: list[VariableDefinition] = []

    @field_validator("name")
    @classmethod
    def validate_name(cls, v: str) -> str:
        v = v.strip()
        if not v:
            raise ValueError("Name cannot be empty")
        return v

    @field_validator("content")
    @classmethod
    def validate_content(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("Content cannot be empty")
        return v


class TemplateUpdate(BaseModel):
    name: str | None = None
    category: TemplateCategory | None = None
    description: str | None = None
    content: str | None = None
    variables: list[VariableDefinition] | None = None


class TemplateResponse(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    name: str
    category: TemplateCategory
    description: str | None
    content: str
    variables: list[VariableDefinition]
    created_by: uuid.UUID | None
    created_at: datetime
    updated_at: datetime


class TemplateListResponse(BaseModel):
    items: list[TemplateResponse]
    total: int


class AdHocPreviewRequest(BaseModel):
    content: str
    variable_values: dict[str, str] = {}


class TemplatePreviewRequest(BaseModel):
    variable_values: dict[str, str] = {}


class PreviewResponse(BaseModel):
    rendered: str
    errors: list[str]
    variables_found: list[str] = []
